pkg/storage/etcd3/v3_helper_test.go

```diff
@@ -0,0 +1,391 @@
+				t.Errorf("#%d: pod want=%#v, get=%#v", i, wantPod, getPod)
+			}
+		}
+	}
+}
+
+func TestV3List(t *testing.T) {
+	cluster := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 1})
+	defer cluster.Terminate(t)
+	helper := newV3Helper(cluster.RandClient(), testapi.Default.Codec(), "")
+	ctx := context.Background()
+
+	// Setup storage with the following structure:
+	//  /
+	//   - one-level/
+	//  |            - test
+	//  |
+	//   - two-level/
+	//               - 1/
+	//              |   - test
+	//              |
+	//               - 2/
+	//                  - test
+	preset := []struct {
+		key       string
+		obj       *api.Pod
+		storedObj *api.Pod
+	}{{
+		key: "/one-level/test",
+		obj: &api.Pod{ObjectMeta: api.ObjectMeta{Name: "foo"}},
+	}, {
+		key: "/two-level/1/test",
+		obj: &api.Pod{ObjectMeta: api.ObjectMeta{Name: "foo"}},
+	}, {
+		key: "/two-level/2/test",
+		obj: &api.Pod{ObjectMeta: api.ObjectMeta{Name: "bar"}},
+	}}
+
+	for i, ps := range preset {
+		preset[i].storedObj = &api.Pod{}
+		err := helper.Set(ctx, ps.key, ps.obj, preset[i].storedObj, 0)
+		if err != nil {
+			t.Fatalf("Set failed: %v", err)
+		}
+	}
+
+	tests := []struct {
+		prefix      string
+		filter      storage.FilterFunc
+		expectedOut []*api.Pod
+	}{{ // test List on existing key
+		prefix:      "/one-level/",
+		filter:      storage.Everything,
+		expectedOut: []*api.Pod{preset[0].storedObj},
+	}, { // test List on non-existing key
+		prefix:      "/non-existing/",
+		filter:      storage.Everything,
+		expectedOut: nil,
+	}, { // test List with filter
+		prefix: "/one-level/",
+		filter: func(obj runtime.Object) bool {
+			pod, ok := obj.(*api.Pod)
+			if !ok {
+				t.Fatal("It should be able to convert obj to *api.Pod")
+			}
+			return pod.Name != preset[0].storedObj.Name
+		},
+		expectedOut: nil,
+	}, { // test List with multiple levels of directories and expect flattened result
+		prefix:      "/two-level/",
+		filter:      storage.Everything,
+		expectedOut: []*api.Pod{preset[1].storedObj, preset[2].storedObj},
+	}}
+
+	for i, tt := range tests {
+		out := &api.PodList{}
+		err := helper.List(ctx, tt.prefix, "0", tt.filter, out)
+		if err != nil {
+			t.Fatalf("List failed: %v", err)
+		}
+		if len(tt.expectedOut) != len(out.Items) {
+			t.Errorf("#%d: length of list want=%d, get=%d", i, len(tt.expectedOut), len(out.Items))
+			continue
+		}
+		for j, wantPod := range tt.expectedOut {
+			getPod := &out.Items[j]
+			if !reflect.DeepEqual(wantPod, getPod) {
+				t.Errorf("#%d: pod want=%#v, get=%#v", i, wantPod, getPod)
+			}
+		}
+	}
+}
+
+func testSetupStore(t *testing.T, etcdClient *clientv3.Client, obj *api.Pod) (context.Context, *v3Helper, string, *api.Pod) {
+	helper := newV3Helper(etcdClient, testapi.Default.Codec(), "")
+	ctx := context.Background()
+	// Setup store with a key and grab the output for returning.
+	key := "/testkey"
+	setOutput := &api.Pod{}
+	err := helper.Set(ctx, key, obj, setOutput, 0)
```

Comment:
I'm not convinced that testSetupStore should be propagating store. This probably should be a separate function.
Then this function should also create the etcd cluster (and return it for termination in the main method).

The you have a clear separate on:
- setup (which is basically the same in all tests
- propagation of store (which depends on the test)